aten/src/ATen/native/cuda/UpSample.cuh

```diff
@@ -0,0 +1,254 @@
+    check_dim_size(grad_output, 4, 3, output_width);
+  }
+}
+
+static inline void upsample_3d_shape_check(
+    const Tensor& input,
+    const Tensor& grad_output,
+    int64_t nbatch,
+    int64_t nchannels,
+    int64_t input_depth,
+    int64_t input_height,
+    int64_t input_width,
+    int64_t output_depth,
+    int64_t output_height,
+    int64_t output_width) {
+  AT_CHECK(
+      input_depth > 0 && input_height > 0 && input_width > 0 &&
+          output_depth > 0 && output_height > 0 && output_width > 0,
+      "Input and output sizes should be greater than 0, but got input (D: ",
+      input_depth,
+      ", H: ",
+      input_height,
+      ", W: ",
+      input_width,
+      ") output (D: ",
+      output_depth,
+      ", H: ",
+      output_height,
+      ", W: ",
+      output_width,
+      ")");
+
+  if (input.defined()) {
+    AT_CHECK(
+        input.numel() != 0 && input.dim() == 5,
+        "Non-empty 5D data tensor expected but got a tensor with sizes ",
+        input.sizes());
+  } else if (grad_output.defined()) {
+    check_dim_size(grad_output, 5, 0, nbatch);
+    check_dim_size(grad_output, 5, 1, nchannels);
+    check_dim_size(grad_output, 5, 2, output_depth);
+    check_dim_size(grad_output, 5, 3, output_height);
+    check_dim_size(grad_output, 5, 4, output_width);
+  }
+}
+
+template <typename accscalar_t>
+__host__ __forceinline__ static accscalar_t area_pixel_compute_scale(
+    int64_t input_size,
+    int64_t output_size,
+    bool align_corners) {
+  if (output_size > 1) {
+    return align_corners ? (accscalar_t)(input_size - 1) / (output_size - 1)
+                         : (accscalar_t)input_size / output_size;
+  } else {
+    return static_cast<accscalar_t>(0);
+  }
+}
+
+template <typename accscalar_t>
+__device__ __forceinline__ static accscalar_t area_pixel_compute_source_index(
+    accscalar_t scale,
+    int64_t dst_index,
+    bool align_corners,
+    bool cubic) {
+  if (align_corners) {
+    return scale * dst_index;
+  } else {
+    accscalar_t src_idx = scale * (dst_index + static_cast<accscalar_t>(0.5)) -
+        static_cast<accscalar_t>(0.5);
+    // See Note[Follow Opencv resize logic]
+    return (!cubic && src_idx < static_cast<accscalar_t>(0))
+        ? static_cast<accscalar_t>(0)
+        : src_idx;
+  }
+}
+
+__device__ __forceinline__ static int64_t nearest_neighbor_compute_source_index(
+    const float scale,
+    int64_t dst_index,
+    int64_t input_size) {
+  const int64_t src_index = min<int64_t>(
+      static_cast<int64_t>(floorf(dst_index * scale)), input_size - 1);
+  return src_index;
+}
+
+/* just affect UpSampleBicubic2d.cu */
+/* TODO: change width and height order in the arguments */
+/* TODO: maybe change x and y order in the arguments */
+/* TODO: maybe change channel and batch order in the arguments */
+template <typename scalar_t>
+__device__ __forceinline__ static scalar_t upsample_get_value_bounded(
+    const PackedTensorAccessor<scalar_t, 4> data,
```

Comment:
I don't think you want to do this. A PackedTensorAccessor is a big struct which contains all of the information necessary so that it can be transmitted across a CUDA kernel call. You may doing an unnecessary copy of all that information with a function signature as done here. At the point you're calling this helper function, you should have decayed the PackedTensorAccessor into a TensorAccessor.

Reply:
More concretely, the suggestion is that this should just take a const reference to a TensorAccessor (or, if you are just looking for the minimal change, I suppose const reference to PackedTensorAccessor would be fine.)

Reply:
Actually, I think this would exactly explain why you were using so many registers. @skrah would you agree with me?

Reply:
There are probably some wiki updates in order here.

Reply:
@ezyang Yes, I just checked and using references to `PackedTensorAccessor` gets down the register count from `124` to `80`.

This is when you do:

```
typedef double scalar_t;
typedef double accscalar_t;
```

Still more than the original, but much better.

Reply:
I think the rest of the difference can be explained by the fact that `THCDeviceTensor` defaults to 32-bit indexing where as `PackedTensorAccessor` defaults to 64-bit indexing. I think for the ports, I'd prefer it if we kept the old indexing integer size, so as to preserve both performance characteristics and bugs. It's probably worth working out if upsample works with larger than 8G tensors; it probably doesn't, much like the rest of the code in THC. cc @colesbury 

This also needs a wiki update.